### PR TITLE
Allow Python interoperability in multilocale settings

### PR DESCRIPTION
**Pull request by lydia-duncan**
When compiling a multilocale library for python, add an argument
to chpl_setup for the number of locales to use.  Removes the
error message for this setting combination.  Stub out an
implementation of cleanupOpaqueArray for the client runtime -
when we support the interoperability notion of opaque arrays, this
function will dispatch to the implementation in the internal modules,
but until that time the function will just generate an error message
and exit.

This moves the definition of a helper function and declares it in
library.h, instead of having it be static to codegen/symbol.cpp.

This change also adds a bunch of tests to ensure it works
(currently only focused on numlocales = 1).  Covers:
- functions that take or return bools
- renaming the python module
- explicitly naming the same module name
- functions that take or return complex vars (not supported yet)
- functions with default values for their arguments
- functions with default values for their complex var arguments (not supported
  yet, both due to complex vars and due to the default value for complex
  arguments)
- functions with default values that are module-scoped variables (not supported
  yet)
- functions that take or return reals
- functions that take or return c_strings
- functions that take or return uints

These all behave as expected (yay!)

I also wrote a test for when the default value of an argument is a prior
argument, but that caused a hang in testing due to issues with cleaning up the
server (so I'm not committing it just yet).  The test that runs with more than 1
locale will sporadically encounter a `GASNET_ERR_BARRIER_MISMATCH` error
(as does a C program, the issue seems to be during the `chpl_library_finalize`
call), so it also isn't committed.  Resolving the latter is one of my next steps.

TODO:
- [ ] run testing on Linux

### CODE REVIEW

File: runtime/etc/src/mli/client_runtime.c [hunk 1]
```diff
@@ -183,5 +186,13 @@ void chpl_library_finalize(void) {
   return;
 }
 
+// TODO: proper implementation once opaque arrays are supported in multilocale
+// interoperability.
+void cleanupOpaqueArray(chpl_opaque_array * arr) {
+  printf("Error: opaque arrays not supported for multilocale libraries");
```
Comment (dlongnecke-cray) [inline runtime/etc/src/mli/client_runtime.c]: I would prefer a `chpl_mli_sdebugf` here just so we know the client was emitting the error in debug mode.

File: test/interop/python/multilocale/boolFunctions.chpl [hunk 2]
```diff
@@ -0,0 +1,15 @@
+/* Tests accepting a bool */
+export proc takesBool(x: bool) {
+  writeln(x);
+}
+
+/* Tests returning a bool */
+export proc getBool(): bool {
```
Comment (dlongnecke-cray) [inline test/interop/python/multilocale/boolFunctions.chpl]: What C type does `bool` map to?

File: test/interop/python/multilocale/checkDiffModname.chpl [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+export proc parallelProg(numTasksToRun: int) {
+  coforall i in 1..numTasksToRun {
+    writeln("In task " + i);
+  }
+}
+
+export proc threadring(passes: int, tasks: int) {
+  var mailbox$: [1..tasks] sync int;
+  mailbox$[1] = 0;
```
Comment (dlongnecke-cray) [inline test/interop/python/multilocale/checkDiffModname.chpl]: Today I learned that Chapel supports closures?!

File: runtime/etc/src/mli/client_runtime.c (hunk 1 above)
Comment (lydia-duncan) [reply runtime/etc/src/mli/client_runtime.c]: Wouldn't that prevent the error from turning up when `--library-ml-debug` is not thrown?

File: test/interop/python/multilocale/boolFunctions.chpl (hunk 2 above)
Comment (lydia-duncan) [reply test/interop/python/multilocale/boolFunctions.chpl]: `chpl_bool`, and that type accepts `bool` from `stdbool.h`

**Pull request merged**